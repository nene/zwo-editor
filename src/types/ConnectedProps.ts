type AnyFn = (...args: any) => any;

type ReturnVoid<F extends AnyFn> = (...args: Parameters<F>) => void;

type FunctionMap = { [key: string]: AnyFn };

type VoidFunctionMap<T extends FunctionMap> = {
  [P in keyof T]: ReturnVoid<T[P]>;
};

// Derives Props type from mapStateToProps and mapDispatchToProps
export type ConnectedProps<
  MapStateToProps extends AnyFn,
  MapDispatchToProps extends FunctionMap = {}
> = ReturnType<MapStateToProps> & VoidFunctionMap<MapDispatchToProps>;
