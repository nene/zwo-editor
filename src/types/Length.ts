export class Duration {
  constructor(public readonly seconds: number) {}
}

export class Distance {
  constructor(public readonly meters: number) {}
}

export type Length = Distance | Duration;

// Alternative approach
//
// This way we could avoid storing class instances to Redux,
// which eliminates problems with serialization & deserialization.
/*
export type Distance = { meters: number };
export type Duration = { seconds: number };
export type Length = Distance | Duration;

export const Distance = (meters: number) => ({ meters });
export const Duration = (seconds: number) => ({ seconds });

export const isDistance = (obj: Length): obj is Distance => 'meters' in obj;
export const isDuration = (obj: Length): obj is Duration => 'seconds' in obj;
*/
