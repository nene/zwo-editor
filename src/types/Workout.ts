import { SportType } from "./SportType";
import { Interval } from "./Interval";
import { LengthType } from "./LengthType";

export interface Workout {
  author: string;
  name: string;
  description: string;
  sportType: SportType;
  lengthType: LengthType;
  tags: string[];
  intervals: Array<Interval>;
}

export function createEmptyWorkout(
  sportType: SportType,
  lengthType: LengthType
): Workout {
  return {
    author: "",
    name: "",
    description: "",
    sportType: sportType,
    lengthType: lengthType,
    tags: [],
    intervals: [],
  };
}
