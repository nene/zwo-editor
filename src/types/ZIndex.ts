export enum ZIndex {
  fader = 1,
  selected = 5,
  instruction = 11,
  tooltip = 100,
  popup = 1000,
}
