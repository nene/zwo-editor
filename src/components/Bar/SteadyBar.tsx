import React, { useState, useEffect } from "react";
import styled from "styled-components";
import { zoneColor, Zones } from "../../types/Zones";
import { Resizable } from "re-resizable";
import Tooltip from "../Tooltip/Tooltip";
import { SteadyInterval } from "../../types/Interval";
import { intensityMultiplier } from "./multipliers";
import { WorkoutMode } from "../../modes/WorkoutMode";

interface SteadyBarProps {
  interval: SteadyInterval;
  mode: WorkoutMode;
  onChange: (interval: SteadyInterval) => void;
  onClick: (id: string) => void;
  selected: boolean;
  showTooltip: boolean;
}

const SteadyBar = ({ interval, mode, ...props }: SteadyBarProps) => {
  const style = { backgroundColor: zoneColor(interval.intensity) };

  const [width, setWidth] = useState(mode.lengthToWidth(interval.length));

  const [height, setHeight] = useState(
    mode.intensityToHeight(interval.intensity)
  );

  const [showTooltip, setShowTooltip] = useState(false);

  const [selected, setSelected] = useState(props.selected);

  useEffect(() => {
    setSelected(props.selected);
  }, [props.selected]);

  const handleCadenceChange = (cadence: number | undefined) => {
    props.onChange({ ...interval, cadence });
  };

  const handleResizeStop = (dWidth: number, dHeight: number) => {
    setWidth(width + dWidth);
    setHeight(height + dHeight);

    notifyChange(dWidth, dHeight);
  };

  const notifyChange = (dWidth: number, dHeight: number) => {
    props.onChange({
      ...interval,
      length: mode.widthToLength(width + dWidth),
      intensity: mode.heightToIntensity(height + dHeight),
    });
  };

  return (
    <Container
      onMouseEnter={() => setShowTooltip(true)}
      onMouseLeave={() => setShowTooltip(false)}
      onClick={() => props.onClick(interval.id)}
      style={props.selected ? { zIndex: 10 } : {}}
    >
      {(selected || showTooltip) && props.showTooltip && (
        <Tooltip
          interval={interval}
          mode={mode}
          onCadenceChange={(cadence) => handleCadenceChange(cadence)}
        />
      )}
      <RoundedResizable
        size={{
          width: mode.lengthToWidth(interval.length),
          height: mode.intensityToHeight(interval.intensity),
        }}
        minWidth={3}
        minHeight={intensityMultiplier * Zones.Z1.min}
        maxHeight={intensityMultiplier * Zones.Z6.max}
        enable={{ top: true, right: true }}
        grid={[1, 1]}
        onResizeStop={(e, direction, ref, d) =>
          handleResizeStop(d.width, d.height)
        }
        onResize={(e, direction, ref, d) => notifyChange(d.width, d.height)}
        style={style}
      ></RoundedResizable>
    </Container>
  );
};

const Container = styled.div`
  position: relative;
`;

const RoundedResizable = styled(Resizable)`
  border-radius: 10px;
  border: 1px solid white;
`;

export default SteadyBar;
