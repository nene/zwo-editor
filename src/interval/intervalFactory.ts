import { v4 as uuidv4 } from "uuid";
import { PaceType } from "../types/PaceType";
import {
  FreeInterval,
  Interval,
  RampInterval,
  RepetitionInterval,
  SteadyInterval,
} from "../types/Interval";
import { Distance, Duration, Length } from "../types/Length";
import { WorkoutMode } from "../modes/WorkoutMode";

const defaultDuration = Duration(300);
const defaultDistance = Distance(1000);
const defaultIntensity = 1.0;
const defaultPace = PaceType.oneMile;

const defaultLength = (mode: WorkoutMode): Length =>
  mode.lengthType === "time" ? defaultDuration : defaultDistance;

export default {
  steady(interval: Partial<SteadyInterval>, mode: WorkoutMode): SteadyInterval {
    return {
      type: "steady",
      id: uuidv4(),
      length: defaultLength(mode),
      intensity: defaultIntensity,
      cadence: undefined,
      pace: defaultPace,
      instructions: [],
      ...interval,
    };
  },

  ramp(interval: Partial<RampInterval>, mode: WorkoutMode): RampInterval {
    return {
      type: "ramp",
      id: uuidv4(),
      length: defaultLength(mode),
      startIntensity: defaultIntensity / 2,
      endIntensity: defaultIntensity,
      cadence: undefined,
      pace: defaultPace,
      instructions: [],
      ...interval,
    };
  },

  free(interval: Partial<FreeInterval>, mode: WorkoutMode): FreeInterval {
    return {
      type: "free",
      id: uuidv4(),
      length: defaultLength(mode),
      cadence: undefined,
      instructions: [],
      ...interval,
    };
  },

  repetition(
    interval: Partial<RepetitionInterval>,
    mode: WorkoutMode
  ): RepetitionInterval {
    return {
      type: "repetition",
      id: uuidv4(),
      repeat: 3,
      onLength: mode.lengthType === "time" ? Duration(30) : Distance(200),
      offLength: mode.lengthType === "time" ? Duration(120) : Distance(200),
      onIntensity: defaultIntensity,
      offIntensity: defaultIntensity / 2,
      onCadence: undefined,
      offCadence: undefined,
      pace: defaultPace,
      instructions: [],
      ...interval,
    };
  },

  // Creates a copy of the interval, with a new ID
  clone(interval: Interval): Interval {
    return { ...interval, id: uuidv4() };
  },
};
