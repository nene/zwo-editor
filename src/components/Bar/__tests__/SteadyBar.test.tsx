import React from "react";
import SteadyBar from "../SteadyBar";
import { Zones } from "../../../types/Zones";
import renderer from "react-test-renderer";
import "@testing-library/jest-dom/extend-expect";
import intervalFactory from "../../../interval/intervalFactory";
import createMode from "../../../modes/createMode";
import { Duration } from "../../../types/Length";

const MockReact = React;

jest.mock("../../Tooltip/Tooltip", () => (props: any) =>
  MockReact.createElement("Tooltip", props)
);

jest.mock("re-resizable", () => ({
  Resizable: (props: any) => MockReact.createElement("Resizable", props),
}));

jest.mock("uuid", () => ({
  v4: () => `mock-id`,
}));

describe("<SteadyBar>", () => {
  it("renders", () => {
    const mode = createMode({
      sportType: "bike",
      ftp: 250,
      weight: 75,
      runningTimes: [],
      lengthType: "time",
    });
    const interval = intervalFactory.steady(
      {
        length: new Duration(50),
        intensity: Zones.Z3.min,
      },
      mode
    );

    const component = renderer.create(
      <SteadyBar
        interval={interval}
        mode={mode}
        selected={false}
        showTooltip={false}
        onChange={() => {}}
        onClick={() => {}}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it("renders without Tooltip when selected, but showTooltip=false", () => {
    const mode = createMode({
      sportType: "bike",
      ftp: 250,
      weight: 75,
      runningTimes: [],
      lengthType: "time",
    });
    const interval = intervalFactory.steady(
      {
        length: new Duration(50),
        intensity: Zones.Z3.min,
      },
      mode
    );

    const component = renderer.create(
      <SteadyBar
        interval={interval}
        mode={mode}
        selected={true}
        showTooltip={false}
        onChange={() => {}}
        onClick={() => {}}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it("renders with Tooltip when selected and showTooltip=true", () => {
    const mode = createMode({
      sportType: "bike",
      ftp: 250,
      weight: 75,
      runningTimes: [],
      lengthType: "time",
    });
    const interval = intervalFactory.steady(
      {
        length: new Duration(50),
        intensity: Zones.Z3.min,
      },
      mode
    );

    const component = renderer.create(
      <SteadyBar
        interval={interval}
        mode={mode}
        selected={true}
        showTooltip={true}
        onChange={() => {}}
        onClick={() => {}}
      />
    );

    expect(component).toMatchSnapshot();
  });
});
