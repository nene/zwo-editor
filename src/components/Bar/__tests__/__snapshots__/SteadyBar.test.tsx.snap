// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SteadyBar> renders 1`] = `
.c0 {
  position: relative;
}

.c1 {
  border-radius: 10px;
  border: 1px solid white;
}

<div
  className="c0"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={Object {}}
>
  <Resizable
    className="c1"
    enable={
      Object {
        "right": true,
        "top": true,
      }
    }
    grid={
      Array [
        1,
        1,
      ]
    }
    maxHeight={500}
    minHeight={25}
    minWidth={3}
    onResize={[Function]}
    onResizeStop={[Function]}
    size={
      Object {
        "height": 190.75,
        "width": 16.666666666666668,
      }
    }
    style={
      Object {
        "backgroundColor": "#00C46A",
      }
    }
  >
    <BarIcons
      instructions={Array []}
    />
  </Resizable>
</div>
`;

exports[`<SteadyBar> renders cadence icon when interval has cadence specified 1`] = `
.c0 {
  position: relative;
}

.c1 {
  border-radius: 10px;
  border: 1px solid white;
}

<div
  className="c0"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={Object {}}
>
  <Resizable
    className="c1"
    enable={
      Object {
        "right": true,
        "top": true,
      }
    }
    grid={
      Array [
        1,
        1,
      ]
    }
    maxHeight={500}
    minHeight={25}
    minWidth={3}
    onResize={[Function]}
    onResizeStop={[Function]}
    size={
      Object {
        "height": 190.75,
        "width": 16.666666666666668,
      }
    }
    style={
      Object {
        "backgroundColor": "#00C46A",
      }
    }
  >
    <BarIcons
      cadence={90}
      instructions={Array []}
    />
  </Resizable>
</div>
`;

exports[`<SteadyBar> renders with Tooltip when selected and showTooltip=true 1`] = `
.c0 {
  position: relative;
}

.c1 {
  border-radius: 10px;
  border: 1px solid white;
}

<div
  className="c0"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "zIndex": 10,
    }
  }
>
  <Tooltip
    interval={
      Object {
        "cadence": undefined,
        "id": "mock-id",
        "instructions": Array [],
        "intensity": 0.763,
        "length": Object {
          "seconds": 50,
        },
        "pace": 0,
        "type": "steady",
      }
    }
    mode={
      BikeMode {
        "ftp": 250,
        "lengthType": "time",
        "sportType": "bike",
        "weight": 75,
      }
    }
    onCadenceChange={[Function]}
  />
  <Resizable
    className="c1"
    enable={
      Object {
        "right": true,
        "top": true,
      }
    }
    grid={
      Array [
        1,
        1,
      ]
    }
    maxHeight={500}
    minHeight={25}
    minWidth={3}
    onResize={[Function]}
    onResizeStop={[Function]}
    size={
      Object {
        "height": 190.75,
        "width": 16.666666666666668,
      }
    }
    style={
      Object {
        "backgroundColor": "#00C46A",
      }
    }
  >
    <BarIcons
      instructions={Array []}
    />
  </Resizable>
</div>
`;

exports[`<SteadyBar> renders without Tooltip when selected, but showTooltip=false 1`] = `
.c0 {
  position: relative;
}

.c1 {
  border-radius: 10px;
  border: 1px solid white;
}

<div
  className="c0"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "zIndex": 10,
    }
  }
>
  <Resizable
    className="c1"
    enable={
      Object {
        "right": true,
        "top": true,
      }
    }
    grid={
      Array [
        1,
        1,
      ]
    }
    maxHeight={500}
    minHeight={25}
    minWidth={3}
    onResize={[Function]}
    onResizeStop={[Function]}
    size={
      Object {
        "height": 190.75,
        "width": 16.666666666666668,
      }
    }
    style={
      Object {
        "backgroundColor": "#00C46A",
      }
    }
  >
    <BarIcons
      instructions={Array []}
    />
  </Resizable>
</div>
`;
